### DISCUSSION TITLE
Install Quantum ESPRESSO on MacOS M1

### DISCUSSION THREAD

**Discussion by nguyen-group** [nguyen-group/QE-SSP#7] [General]
### Install GCC, OpenMPI, wget, ScaLapack via brew:
```
$ brew install gcc open-mpi cmake wget scalapack
```
### Download Quantum ESPRESSO via wget:
```
$ wget https://gitlab.com/QEF/q-e/-/archive/qe-7.2/q-e-qe-7.2.tar.gz
```
### Un-tar the source files:
```
$ tar -zxvf q-e-qe-7.2.tar.gz
```
### Go to the QE directory and configure:
```
$ cd q-e-qe-7.2
$ ./configure
```
### Modify make.inc in the q-e-qe-7.2 directory for macos m1:
**You might need to run ./configure in the q-e-qe-7.2 directory first; then it will create the file make.inc in the same directory.**
```
Change "CPP = cpp" in line 107 of make.inc to "CPP = gcc -E"
```
Note: If we do not change make.inc, we might get the error: 
```
1 warning generated.
clang: error: no input files
make[1]: *** [laxlib.fh] Error 1
make: *** [libla] Error 1
```
### Compile everything:
```
$ make all
```
### Setting Bash variables
```
export OMP_NUM_THREADS=1
export PATH=$HOME/q-e-qe-7.2/bin:$PATH
```
Change $HOME to your path, OMP_NUM_THREADS to set the number of threads to use for parallel.

**Comment by Hary-Lee** [comment]
Hi can you explain this step more specifically!
Change "CPP = cpp" in line 107 of make.inc to "CPP = gcc -E"
Since I typed this command to Terminal, it show this error: zsh: command not found: Change.
Thank you!

**Reply by nguyen-group** [reply]
Dear Hary-Lee,

It is not the command line, but you need to find the file "make.inc" in the QE folder and modify it by changing "CPP = cpp" in line 107 of make.inc to "CPP = gcc -E".

Best,
Nguyen

**Reply by Hary-Lee** [reply]
Dear Mr. Hung,

First of all, thank you for your replying to my previous question. But I'm sorry that I still could't do it successfully. I'm trying to install QE to my Mac book air M2, macOS Sonoma 14.4.1. At first, I think it's quite similar with M1, so that I try to do all the steps of your instruction above. But even when I change "CPP = cpp" in line 107 of make.inc to "CPP = gcc -E", it still noticed the error. 
<img width="452" alt="Ảnh màn hình 2024-05-16 lúc 21 18 03" src="https://github.com/nguyen-group/QE-SSP/assets/166190303/f20c9eed-2fc7-4316-8c3d-5c410a4ac034">
<img width="572" alt="Ảnh màn hình 2024-05-16 lúc 21 41 35" src="https://github.com/nguyen-group/QE-SSP/assets/166190303/3df40852-2fad-4fff-b291-8f68154f9aaf">

Could you help me with this, please?!

Thanks & Best Regards.

**Reply by nguyen-group** [reply]
Hi, 

I am sorry that I don't have a M2 for testing. Thus, I have no idea how to solve this error.
You might probably consider using MateriApps to install QE on Mac M2. 
The information can be found here:
https://ma.issp.u-tokyo.ac.jp/en/app-post/5274

Best,
Nguyen

**Comment by andres-hidalgo4** [comment]
Hi
I am currently trying to install QE on MacOS M1.  I am following the mentioned steps but after chaning the CPP value to gcc -E. I still have the error.
Do you maybe know how to solve it?
*I am using the 7.3 version of quantum espresso
<img width="629" alt="Screenshot 2025-01-27 at 14 53 33" src="https://github.com/user-attachments/assets/01724680-e85b-48d5-8281-ab7a81b76307" />

**Reply by nguyen-group** [reply]
Hi,

I am sorry that I don't know what this error means.
For Mac, another suggestion is using Spack. You can find the information at: https://github.com/nguyen-group/QE-SSP/discussions/9

Best
Nguyen

**Comment by alien2327** [comment]
For anyone trying to install Quantum Espresso (qe) based on clang on macOS...

I was able to build it successfully. However, it was a bit tedious. I cannot guarantee that this is a normal build method in terms of C or Fortran culture or computer calculations, so I will explain it on the assumption that each person is responsible for themselves.

Here is what I did. All builds use clang.
(1) First, install the necessary packages (openmpi, scalapack, libxc, etc.) using Homebrew.
(2) Run configure with the appropriate flags. Note that -fallow-argument-mismatch must be included in FCFLAGS. (As mentioned earlier, this is not a safe method. It is purely focused on achieving a successful build.)
(3) As of version 7.4.1, errors will occur in the header files under LAXlib and KS_Solvers.
(4) Modify the laxlib_*.h files under LAXlib as follows: (i) Copy all h files and change the extension to fh, (ii) Open each fh file and delete all CUDA settings in #if .. #endif. At the time of writing this article, there is no need to use CUDA on mac, so this should be fine. (iii) Do not run make clean; instead, compile directly with make all. The fh files are part of the build, so running make clean will delete the newly created fh files.
(5) Apply the same process to the ks_solver_interfaces.h file under the KS_Solvers directory.
(6) On my computer (Mac Studio M3 Max 512GB), the build was successful. I have not tested this on other Apple Silicon systems, and since I am not the developer of qe, further analysis is difficult.

I hope this serves as a reference for users who wish to try this on Apple Silicon.

This is my setting
./configure MPIF90=mpif90 FFLAGS="-Ofast -fdefault-double-8 -fallow-argument-mismatch -I/opt/homebrew/opt/openblas/include -I/opt/homebrew/opt/openmpi/include -I/opt/homebrew/opt/fftw/include -I/opt/homebrew/include" CC=mpicc CFLAGS="-O3 -I/opt/homebrew/opt/openblas/include -I/opt/homebrew/opt/openmpi/include -I/opt/homebrew/opt/fftw/include -I/opt/homebrew/include" LDFLAGS="-L/opt/homebrew/lib -L/opt/homebrew/opt/openblas/lib -L/opt/homebrew/opt/openmpi/lib -L/opt/homebrew/opt/scalapack/lib -L/opt/homebrew/opt/fftw/lib" --with-libxc=yes --with-libxc-prefix=/opt/homebrew/Cellar/libxc/7.0.0
